[mlir][vector] proper masking support for contract lowering

ftynse opened the pull request:
Support all known permutations when lowering masked vector.contract to vector.outerproduct, and not just the canonical permutation.

ftynse committed:
[mlir][vector] proper masking support for contract lowering

Support all known permutations when lowering masked vector.contract to
vector.outerproduct, and not just the canonical permutation.

nicolasvasilache reviewed (approved):
Nice, thanks for plugging these holes

dcaballe reviewed (commented):
Awesome. Thanks, Alex!

ftynse committed:
fix implementation-defined behavior mismatch

Calling t() constructs IR, so it should not happen from multiple
function arguments that will be executed in different order on Windows
and Linux.
